### ISSUE TITLE
Dynamic import from within an ES module never resolves

### ISSUE THREAD

**Issue by fatcerberus** [Microsoft/ChakraCore#3902]
A dynamic `import()` of a module that wasn't already imported statically will never resolve.  This only affects `import()` done in an ES module--dynamic imports loaded using the FetchImportedModuleFromScript strategy resolve just fine.

Repro:
```js
// test.js
console.log("running test.js");
WScript.LoadScriptFile('./pig.js', 'module');
```
```js
// pig.js
console.log("running pig.js");
console.log("starting to load cow.js");
import("./cow.js").then(module => {
	console.log("successfully loaded cow.mjs");
	console.log(`the cow says: ${module.Cow}`);
}).catch(e => {
	console.log(`error: ${e}`);
});
```
```js
// cow.js
console.log("running cow.js");
export const Cow = "cowz eat kittiez";
```

Output:
```
D:\src\ChakraCore\Build\VcBuild\bin\x64_debug>ch test.js
running test.js
running pig.js
starting to load cow.js
```

By contrast, `import()`ing from global code:
```
D:\src\ChakraCore\Build\VcBuild\bin\x64_debug>ch pig.js
running pig.js
starting to load cow.js
running cow.js
successfully loaded cow.mjs
the cow says: cowz eat kittiez
```

**Comment by fatcerberus** [comment]
Having done some debugging for this problem already, it seems that ChakraCore never calls the NotifyModuleReady callback for dynamic imports unless their module records are explicitly initialized as top-level (`referencingModule` is nullptr).

My original workaround for this bug was to create all module records as roots regardless of heritage, however that proved to be problematic as imports would randomly fail to resolve.

**Comment by fatcerberus** [comment]
Follow-up on my repro above, if one changes pig.js to add a static import of the same module:
```js
import { Cow } from './cow.js'

console.log("running pig.js");
console.log("starting to load cow.js");
import("./cow.js").then(module => {
	console.log("successfully loaded cow.js");
	console.log(`the cow says: ${module.Cow}`);
}).catch(e => {
	console.log(`error: ${e}`);
});
```

Then everything is right with the world:
```
D:\src\ChakraCore\Build\VcBuild\bin\x64_debug>ch test.js
running test.js
running cow.js
running pig.js
starting to load cow.js
successfully loaded cow.js
the cow says: cowz eat kittiez
```

Of course having to declare a static import for all dynamic imports kind of defeats the purpose of dynamic import... 😉

**Comment by rhuanjl** [comment]
OK found a fix for this, the issue is that currently the only time chakra tells the host that a module is ready is when a parent module is ready but this condition should never trigger for a dynamic module.

The call to tell the host a module is ready is wrapped in a condition on line 310 of SourceTextMOduleRecord.cpp
https://github.com/Microsoft/ChakraCore/blob/master/lib/Runtime/Language/SourceTextModuleRecord.cpp#L310

Condition is:
```cc
if (!WasDeclarationInitialized() && isRootModule)
```
If we revise that to:
```cc
if (!WasDeclarationInitialized() && (isRootModule || this->promise != nullptr))
```
It will catch dynamic imports as well as parent modules - testing this with the above test case and ch results in the dynamic import now working.

**Comment by liminzhu** [comment]
FYI @akroshg

**Comment by rhuanjl** [comment]
@leirocks directing my questions to you as you've been assigned to this issue.

So between @fatcerberus and I we have:
1. identified an issue,
2. written a test case that makes it repro in ch AND
3. found a fix - I can further explain the fix if wanted - it very clearly seems to me to be correcting an accidental omission.

I note that per the contribution guidelines external pull requests are only accepted for Approved Bugs and Features - I have not therefore submitted a pull request as at the moment this issue has not been tagged as either of those.

I think this should count as a Bug fix but this issue is not currently tagged as a bug - what do we need to do to get it tagged as a bug and get a pull request to fix it accepted? If further help from me would be useful please let me know - conversely if you'd rather submit this fix on your own go ahead.

**Comment by akroshg** [comment]
@rhuanjl thanks for reaching out. As you have said this is indeed a bug and it should be tagged as that.
Your suggestive fix seems to be correct and we will be happy to take your change.
@fatcerberus thanks for raising this issue. And of course you shouldn't have to put static import before dynamic import, not only it defeats the purpose but the output will be wrong as well. As you see that the message ```running cow.js``` should come after ```running pig.js```.

**Comment by rhuanjl** [comment]
Not the quick fix I'd hoped - I'm still hoping to fix this but my suggestion above and current PR are not the answer.

I'm hoping to solve it and update the PR but not sure how long it will take, if anyone else seeing this has the answer I'll be very happy to step back.

**Original problem:**
1. ChakraCore does not want to notify the host that a module is ready until all dependencies are also ready - so notification is done for the root module when all children have notified
2. If a module is dynamically imported from within a static module it is treated as a child of that static module
3. BUT the host has already been notified that that static module is ready
 => the host is never notified after the dynamic module is imported

**My proposed fix**
Notify the host for all dynamic modules this fixes the issue for a dynamic import from within a static import.

**Why it doesn't work**
The fix breaks dynamic imports done from scripts (not modules); the host is notified that they are ready BEFORE the callback is set this results in a segfault (or a failed Assert with the Debug build).

I'm struggling to work out why this is - it seems that the dynamic loaded from a module is perhaps able to take advantage of the callbacks set for its parent which obviously doesn't work for a dynamic module loaded from script - I've tried pushing the call later and later in the process but can't find a safe place for it - there appears to be something I'm missing on how this works :(.

**Comment by rhuanjl** [comment]
Minor updates but not a solution yet:
1. see edit to above post the actual error I'm trying to solve isn't quite I'd posted earlier.
2. I haven't mentioned this before but possibly worth highlighting, inside the function "PostProcessDynamicModuleImport" (PPDMI) there is an alternate call that could notify the host that the module is ready rather than using the one at line 310 in PrepareForModuleDeclarationInitialization, BUT the call in PPDMI is guarded with a condition of "this->ParentsNotified()" and on all tests I've managed to do all calls to NotifyParentsAsNeeded() come later in the call chain and so currently this appears to be dead code - I further note that if I change this condition it results in the same failure of the test bug_issue_3257 and other errors too. I've also tried re-calling PostProcessDynamicModuleImport later in the process but the spots I've tried putting it so far still fails bug_issue_3257.
